File: SSUMobile/SSUAppDelegate.m
//
//  SSUAppDelegate.m
//  SSUMobile
//
//  Created by Andrew Huss on 1/24/13.
//  Copyright (c) 2013 Sonoma State University Department of Computer Science. All rights reserved.
//

@import Foundation;
@import CoreSpotlight;

#import "SSUAppDelegate.h"
#import "SSUSettingsConstants.h"
#import "SSURadioStreamer.h"
#import "SSUModuleBase.h"
#import "SSULogging.h"
#import "SSUConfiguration.h"
#import "SSUModuleServices.h"
#import "SSUSpotlightServices.h"

#import <SDWebImage/SDImageCache.h>

@interface SSUAppDelegate()

@end

@implementation SSUAppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    [SSULogging setupLogging];
    [self setupConfiguration];
    [self setupStyles];
    
    SSULogDebug(@"%@", SSUDocumentsDirectory());
    SSULogDebug(@"%@", SSUCachesDirectory());
    
    if (![[NSFileManager defaultManager] fileExistsAtPath:SSUApplicationSupportDirectory().path]) {
        [[NSFileManager defaultManager] createDirectoryAtURL:SSUApplicationSupportDirectory()
                                 withIntermediateDirectories:YES
                                                  attributes:nil
                                                       error:nil];
    }
    
    //Update all modules
    if ([self isFirstLaunchForCurrentVersion]) {
        SSULogDebug(@"First launch");
        [self clearLocalDatabases];
        [[SSUConfiguration sharedInstance] setBool:NO forKey:[self firstLaunchKey]];
        [[SSUModuleServices sharedInstance] setupAll];
        [self showWelcomeMessage];
    }
    else {
        [[SSUModuleServices sharedInstance] setupAll];
    }
    
    [[SDImageCache sharedImageCache] setMaxCacheSize:1024*1000*100]; // 100MB max cache size
    
    return YES;
}

- (void) applicationDidBecomeActive:(UIApplication *)application {
    // Load settings from moonlight
    if (![self isFirstLaunchForCurrentVersion]) {
        [self loadRemoteConfiguration];
        [[SSUModuleServices sharedInstance] updateAll];
    }
}

- (void) setupStyles {
    
    UIColor * mainColor = SSU_BLUE_COLOR;
    [[UINavigationBar appearance] setTintColor:[UIColor whiteColor]];
    [[UINavigationBar appearance] setBarTintColor:mainColor];
    [[UINavigationBar appearance] setBarStyle:UIBarStyleBlack];
    [[UIToolbar appearance] setBarStyle:UIBarStyleBlack];
    [[UIToolbar appearance] setBarTintColor:mainColor];
    [[UIToolbar appearance] setTintColor:[UIColor whiteColor]];
    [[UISearchBar appearance] setBarTintColor:mainColor];
    [[UISearchBar appearance] setTintColor:mainColor];
    [[UISearchBar appearance] setBackgroundColor:mainColor];
    [[UISearchBar appearance] setBarStyle:UIBarStyleBlack];
    [[UISegmentedControl appearance] setTintColor:mainColor];
    [[UISegmentedControl appearanceWhenContainedIn:[UIToolbar class], nil] setTintColor:[UIColor whiteColor]];
    [[UISegmentedControl appearanceWhenContainedIn:[UINavigationBar class], nil] setTintColor:[UIColor whiteColor]];
    [[UIBarButtonItem appearanceWhenContainedIn:[UIToolbar class], nil] setTintColor:[UIColor whiteColor]];

}

- (void) setupConfiguration {
    // Defaults not present in the JSON files
    NSDictionary *userDefaults = @{
                                   SSUDirectorySortOrderKey: @(kSSUFirstLast),
                                   SSUDirectoryDisplayOrderKey: @(kSSUFirstLast),
                                   [self firstLaunchKey] : @(YES)
                                   };
    
    [[SSUConfiguration sharedInstance] registerDefaults:userDefaults];
    // Load JSON defaults included in app bundle
    NSString * path = [[NSBundle mainBundle] pathForResource:@"defaults.json" ofType:nil];
    [[SSUConfiguration sharedInstance] loadDefaultsFromFilePath:path];
}

- (void) loadRemoteConfiguration {
    NSURL * configURL = [NSURL URLWithString:[SSUMoonlightBaseURL stringByAppendingPathComponent:@"settings"]];
    NSArray * classes = [[SSUConfiguration sharedInstance] stringArrayForKey:SSUModulesEnabledKey];
    [[SSUConfiguration sharedInstance] loadFromURL:configURL completion:^(NSError *error) {
        SSULogDebug(@"After loading from moonlight: %@", [[SSUConfiguration sharedInstance] dictionaryRepresentation]);
        if (![classes isEqualToArray:[[SSUConfiguration sharedInstance] stringArrayForKey:SSUModulesEnabledKey]]) {
            [[SSUModuleServices sharedInstance] loadModules];
        }
    }];
}

- (NSString *) firstLaunchKey {
    NSString * const versionKey = @"CFBundleShortVersionString";
    NSString * const appVersion = [[[NSBundle mainBundle] infoDictionary] objectForKey:versionKey];
    NSString * const suffix = [NSString stringWithFormat:@"_%@", appVersion];
    NSString * const key = [SSUAppIsFirstLaunchKey stringByAppendingString:suffix];
    return key;
}

- (BOOL) isFirstLaunchForCurrentVersion {
    return [[SSUConfiguration sharedInstance] boolForKey:[self firstLaunchKey]];
}

/**
 Remote Control Notification
 
 This is called when the user interacts with the media controls on the lock screen
 or in the command center
 */
- (void)remoteControlReceivedWithEvent:(UIEvent *)receivedEvent {
    
    if (receivedEvent.type == UIEventTypeRemoteControl) {
        switch (receivedEvent.subtype) {
            case UIEventSubtypeRemoteControlPause:
            case UIEventSubtypeRemoteControlPlay:
            case UIEventSubtypeRemoteControlTogglePlayPause:
                [[SSURadioStreamer sharedInstance] togglePlayer];
                break;
            case UIEventSubtypeRemoteControlPreviousTrack:
                break;
            case UIEventSubtypeRemoteControlNextTrack:
                break;
            default:
                break;
        }
    }
}

- (void) applicationDidReceiveMemoryWarning:(UIApplication *)application {
    [[SDImageCache sharedImageCache] clearMemory];
}

#pragma mark - NSUserActivity / CoreSpotlight

- (BOOL) application:(UIApplication *)application
continueUserActivity:(nonnull NSUserActivity *)userActivity
  restorationHandler:(nonnull void (^)(NSArray * _Nullable))restorationHandler {
    if ([userActivity.activityType isEqualToString:CSSearchableItemActionType]) {
        NSString * identifier = userActivity.userInfo[CSSearchableItemActivityIdentifier];
        NSArray * spotlightModules = [[SSUModuleServices sharedInstance] modulesConformingToProtocol:@protocol(SSUSpotlightSupportedProtocol)];
        for (id<SSUSpotlightSupportedProtocol> module in spotlightModules) {
            if ([module recognizesIdentifier:identifier]) {
                UIViewController * viewController = [module viewControllerForSearchableItemWithIdentifier:identifier];
                [[NSNotificationCenter defaultCenter] postNotificationName:SSUSpotlightActivityRequestingDisplayNotification object:viewController];
            }
        }
    }
    return NO;
}

- (BOOL) application:(UIApplication *)application willContinueUserActivityWithType:(NSString *)userActivityType {
    return [userActivityType isEqualToString:CSSearchableItemActionType];
}

#pragma mark - Helper

- (void) showWelcomeMessage {
    UIAlertView * alert = [[UIAlertView alloc] initWithTitle:@"Welcome to SSUMobile!"
                                                     message:@"Welcome to the lastest version of SSUMobile! In order to provide you with up to date information, some data will be download from internet."
                                                    delegate:self
                                           cancelButtonTitle:@"Okay"
                                           otherButtonTitles:nil];
    [alert show];
}

- (void) alertView:(UIAlertView *)alertView didDismissWithButtonIndex:(NSInteger)buttonIndex {
    [[SSUModuleServices sharedInstance] updateAll];
    [self loadRemoteConfiguration];
}

/**
 Removes any existing Core Data database files stored in the documents directory.
 */
- (void) clearLocalDatabases {
    NSString * extensionToDelete = @".sqlite";
    [self deleteFilesInDiretory:SSUDocumentsDirectory() matchingExtension:extensionToDelete];
    [self deleteFilesInDiretory:SSUApplicationSupportDirectory() matchingExtension:extensionToDelete];
}

- (void) deleteFilesInDiretory:(NSURL *)directory matchingExtension:(NSString *)extension {
    BOOL isDirectory = NO;
    BOOL exists = [[NSFileManager defaultManager] fileExistsAtPath:directory.path isDirectory:&isDirectory];
    if (!(exists && isDirectory)) {
        SSULogError(@"Attempting to remove files from a path which does not exist or is not a directory: %@", directory);
        return;
    }
    
    NSArray * filePaths = [[NSFileManager defaultManager] contentsOfDirectoryAtURL:directory
                                                        includingPropertiesForKeys:nil
                                                                           options:NSDirectoryEnumerationSkipsHiddenFiles
                                                                             error:nil];
    for (NSURL * path in filePaths) {
        if ([path.path rangeOfString:extension].location == NSNotFound) {
            continue;
        }
        SSULogDebug(@"Removing file at path: %@",path);
        [[NSFileManager defaultManager] removeItemAtURL:path error:nil];
    }
}

@end

File: SSUMobile/SSUAppDelegate.swift
//
//  SSUAppDelegate.swift
//  SSUMobile
//
//  Created by Eric Amorde on 4/7/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import UIKit
import MBProgressHUD
import Mockingjay

@UIApplicationMain
class SSUAppDelegate: UIResponder, UIApplicationDelegate {
    
    var window: UIWindow?
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey : Any]? = nil) -> Bool {
        
        SSULogging.setupLogging()
        SSULogging.logDebug("\(SSUDocumentsDirectory())")
        SSULogging.logDebug("\(SSUCachesDirectory())")
        
        setupConfiguration()
        setupStyles()
        
        SSUModuleServices.instance.loadModules()
        
        if !FileManager.default.fileExists(atPath: SSUApplicationSupportDirectory().path) {
            try? FileManager.default.createDirectory(at: SSUApplicationSupportDirectory(), withIntermediateDirectories: true, attributes: nil)
        }
        
        if isFirstLaunchForCurrentVersion() {
            SSULogging.logDebug("First launch")
            clearLocalDatabases()
            SSUModuleServices.instance.setupAll()
            showWelcomeMessage(completion: {
                SSUConfiguration.sharedInstance().setBool(false, forKey: self.firstLaunchKey())
                SSUModuleServices.instance.updateAll()
                self.loadRemoteConfiguration()
            })
        } else {
            SSUModuleServices.instance.setupAll()
        }
        
        return true
    }
    
    func applicationDidBecomeActive(_ application: UIApplication) {
        // Load settings from moonlight
        if !isFirstLaunchForCurrentVersion() {
            loadRemoteConfiguration()
            SSUModuleServices.instance.updateAll()
        }
    }
    
    func setupStyles() {
        let mainColor = UIColor.ssuBlue
        UINavigationBar.appearance().tintColor = .white
        UINavigationBar.appearance().barTintColor = mainColor
        UINavigationBar.appearance().barStyle = .black
        UIToolbar.appearance().barStyle = .black
        UIToolbar.appearance().barTintColor = mainColor
        UIToolbar.appearance().tintColor = .white
        UISearchBar.appearance().barTintColor = mainColor
        UISearchBar.appearance().tintColor = mainColor
        UISearchBar.appearance().backgroundColor = mainColor
        UISearchBar.appearance().barStyle = .black
        UISegmentedControl.appearance().tintColor = mainColor
        if #available(iOS 9.0, *) {
            UISegmentedControl.appearance(whenContainedInInstancesOf: [UIToolbar.self]).tintColor = .white
            UISegmentedControl.appearance(whenContainedInInstancesOf: [UINavigationBar.self]).tintColor = .white
            UIBarButtonItem.appearance(whenContainedInInstancesOf: [UIToolbar.self]).tintColor = .white
        }
        
        MBProgressHUD.appearance().contentColor = mainColor
    }
    
    func setupConfiguration() {
        // Defaults not present in the JSON files
        let userDefaults: [String:Any] = [
            SSUDirectorySortOrderKey: kSSUFirstLast.rawValue,
            SSUDirectoryDisplayOrderKey: kSSUFirstLast.rawValue,
            firstLaunchKey(): true
        ]
        SSUConfiguration.sharedInstance().registerDefaults(userDefaults)
        // Load JSON defaults included in app bundle
        if let path = Bundle.main.path(forResource: "defaults.json", ofType: nil) {
            SSUConfiguration.sharedInstance().loadDefaults(fromFilePath: path)
        }
    }
    
    func loadRemoteConfiguration() {
        // TODO: move this stuff to SSUConfiguration
        let versionKey = "CFBundleShortVersionString"
        guard let appVersion = Bundle.main.infoDictionary?[versionKey] else {
            SSULogging.logError("Unable to retrieve app version - cannot load remote config")
            return
        }
        let endpoint = "ssumobile/settings/ios/\(appVersion)/"
        guard let baseURL = URL(string: SSUMoonlightBaseURL),
            let configURL = URL(string: endpoint, relativeTo: baseURL) else {
            SSULogging.logError("Unable to create url for remote config")
            return
        }
        if SSUDebugUtils.shouldMockConfig {
            let fileURL = Bundle.main.url(forResource: "defaults.json", withExtension: nil)!
            let data = try! Data(contentsOf: fileURL)
            MockingjayProtocol.addStub(matcher: uri(configURL.absoluteString), builder: jsonData(data))
        }
        SSUConfiguration.sharedInstance().load(from: configURL) { (error) in
            if let error = error {
                SSULogging.logError("Error loading config: \(error)")
            } else {
                SSUModuleServices.instance.loadModules()
            }
        }
    }
    
    func firstLaunchKey() -> String {
        let versionKey = "CFBundleShortVersionString"
        guard let appVersion = Bundle.main.infoDictionary?[versionKey] else {
            SSULogging.logError("Unable to retrieve app version")
            return SSUAppIsFirstLaunchKey
        }
        let key = "\(SSUAppIsFirstLaunchKey)_\(appVersion)"
        return key
    }
    
    func isFirstLaunchForCurrentVersion() -> Bool {
        return SSUConfiguration.sharedInstance().bool(forKey: firstLaunchKey())
    }
    
    
    /**
     Remote Control Notification
     
     This is called when the user interacts with the media controls on the lock screen
     or in the command center
     */
    override func remoteControlReceived(with event: UIEvent?) {
        super.remoteControlReceived(with: event)
        guard let receivedEvent = event else {
            return
        }
        if receivedEvent.type == .remoteControl {
            switch (receivedEvent.subtype) {
            case .remoteControlPause,
                 .remoteControlPlay,
                 .remoteControlTogglePlayPause:
                SSURadioStreamer.sharedInstance().togglePlayer()
            default:
                break;
            }
        }
    }
    
    // MARK: - NSUserActivity / CoreSpotlight
    
    func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
        if #available(iOS 9.0, *) {
            if let identifier = userActivity.userInfo?[CSSearchableItemActivityIdentifier] as? String, userActivity.activityType == CSSearchableItemActionType {
                let spotlightModules = SSUModuleServices.instance.modulesConformingTo(type: SSUSpotlightSupportedProtocol.self)
                for module in spotlightModules {
                    if module.recognizesIdentifier(identifier) {
                        let viewController = module.viewControllerForSearchableItem(withIdentifier: identifier)
                        NotificationCenter.default.post(name: NSNotification.Name.SSUSpotlightActivityRequestingDisplay, object: viewController)
                        return true
                    }
                }
            }
        }
        return false
    }
    
    func application(_ application: UIApplication, willContinueUserActivityWithType userActivityType: String) -> Bool {
        if #available(iOS 9.0, *) {
            return userActivityType == CSSearchableItemActionType
        }
        return false
    }
    
    func showWelcomeMessage(completion: (() -> Void)? = nil) {
        let alert = UIAlertController(title: "Welcome to SSUMobile",
                                      message: "Welcome to the lastest version of SSUMobile! In order to provide you with up to date information, some data will be download from internet.",
                                      preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Okay", style: .default, handler: { (action) in
            completion?()
        }))
    }
        
    /**
     Removes any existing Core Data database files stored in the documents directory.
     */
    func clearLocalDatabases() {
        let extensionToDelete = ".sqlite"
        deleteFilesInDirectory(SSUDocumentsDirectory(), matchingExtension: extensionToDelete)
        deleteFilesInDirectory(SSUApplicationSupportDirectory(), matchingExtension: extensionToDelete)
    }
    
    func deleteFilesInDirectory(_ directory: URL?, matchingExtension fileExtension: String) {
        guard let url = directory else { return }
        let exists = FileManager.default.fileExists(atPath: url.path)
        if !exists {
            SSULogging.logError("Directory does not exist: \(url)")
            return
        }
        
        do {
            let filePaths = try FileManager.default.contentsOfDirectory(atPath: url.path)
            for path in filePaths {
                if path.hasSuffix(fileExtension) {
                    try FileManager.default.removeItem(atPath: path)
                }
            }
        } catch {
            SSULogging.logError("Unable to find files in directory \(url)")
            return
        }
    }
    
}
